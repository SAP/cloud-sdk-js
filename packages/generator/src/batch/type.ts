import { StructureKind, TypeAliasDeclarationStructure } from 'ts-morph';
import { VdmServiceMetadata } from '../vdm-types';
import { operationReturnType } from '../operations';

/**
 * @internal
 */
export function readRequestType(
  service: VdmServiceMetadata
): TypeAliasDeclarationStructure {
  return {
    kind: StructureKind.TypeAlias,
    name: `Read${service.className}RequestBuilder<DeSerializersT extends DeSerializers>`,
    isExported: true,
    type: getReadRequestType(service)
  };
}
/**
 * @internal
 */
export function writeRequestType(
  service: VdmServiceMetadata
): TypeAliasDeclarationStructure {
  return {
    kind: StructureKind.TypeAlias,
    name: `Write${service.className}RequestBuilder<DeSerializersT extends DeSerializers>`,
    isExported: true,
    type: getWriteRequestType(service)
  };
}

function getWriteRequestType(service: VdmServiceMetadata): string {
  const createUpdateDeleteBuilderTypes = service.entities.map(
    e =>
      `CreateRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT> | UpdateRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT> | DeleteRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT>`
  );
  const functionImportsReturnTypes = service.functionImports
    .filter(fn => fn.httpMethod.toLowerCase() !== 'get')
    .map(fn => operationReturnType(fn));
  const actionImportsReturnTypes =
    service.actionImports?.map(fn => operationReturnType(fn)) ?? [];
  return [
    ...createUpdateDeleteBuilderTypes,
    ...functionImportsReturnTypes,
    ...actionImportsReturnTypes
  ].join(' | ');
}

function getReadRequestType(service: VdmServiceMetadata): string {
  const getAllBuilderTypes = service.entities.map(
    e => `GetAllRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT>`
  );

  const getByKeyBuilderTypes = service.entities.map(
    e =>
      `GetByKeyRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT>`
  );
  const functionImportsReturnTypes = service.functionImports
    .filter(fn => fn.httpMethod.toLowerCase() === 'get')
    .map(fn => operationReturnType(fn));

  return [
    ...getAllBuilderTypes,
    ...getByKeyBuilderTypes,
    ...functionImportsReturnTypes
  ].join(' | ');
}
