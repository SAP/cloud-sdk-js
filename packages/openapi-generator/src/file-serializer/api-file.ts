import { codeBlock, documentationBlock, unixEOL } from '@sap-cloud-sdk/util';
import {
  serializeImports,
  Import
} from '@sap-cloud-sdk/generator-common/internal';
import {
  OpenApiApi,
  OpenApiOperation,
  OpenApiReferenceSchema
} from '../openapi-types';
import { collectRefs, getUniqueRefs } from '../schema-util';
import { serializeOperation } from './operation';

/**
 * Serialize an API representation to a string representing the resulting API file.
 * @param api - Representation of an API.
 * @param serviceName - Service name for which the API is created.
 * @returns The serialized API file contents.
 * @internal
 */
export function apiFile(api: OpenApiApi, serviceName: string): string {
  const imports = serializeImports(getImports(api));
  const apiDoc = apiDocumentation(api, serviceName);
  const apiContent = codeBlock`
export const ${api.name} = {
  ${api.operations.map(operation => serializeOperation(operation)).join(',\n')}
};
`;

  return [imports, apiDoc, apiContent].join(unixEOL);
}

/**
 * Get the unique reference schemas for all request body types in the given operation list.
 * @param operations - The given operation list.
 * @returns The list of unique referenced body type schemas.
 */
function collectRefsFromOperations(
  operations: OpenApiOperation[]
): OpenApiReferenceSchema[] {
  return getUniqueRefs(
    operations.reduce(
      (referenceTypes, operation) => [
        ...referenceTypes,
        ...collectRefs(
          ...[
            operation.requestBody?.schema,
            operation.response,
            ...operation.queryParameters.map(({ schema }) => schema)
          ]
        )
      ],
      []
    )
  );
}

function getImports(api: OpenApiApi): Import[] {
  const refs = collectRefsFromOperations(api.operations).map(
    requestBodyType => requestBodyType.schemaName
  );

  const refImports = {
    names: refs,
    moduleIdentifier: './schema',
    typeOnly: true
  };
  const openApiImports = {
    names: ['OpenApiRequestBuilder'],
    moduleIdentifier: '@sap-cloud-sdk/openapi'
  };

  return [openApiImports, refImports];
}

/**
 * @internal
 */
export function apiDocumentation(api: OpenApiApi, serviceName: string): string {
  return documentationBlock`
  Representation of the '${api.name}'.
  This API is part of the '${serviceName}' service.`;
}
