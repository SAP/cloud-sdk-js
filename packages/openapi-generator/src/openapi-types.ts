import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  operationName: string;
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

const supportedMethods = {
  get: 'get',
  put: 'put',
  post: 'post',
  patch: 'patch',
  delete: 'delete',
  head: 'head',
  options: 'options'
} as const;

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export type Method = typeof supportedMethods[keyof typeof supportedMethods];

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get supported methods.
 * @returns Methods supported by OpenApi and SAP Cloud SDK.
 */
export function methods(): Method[] {
  return Object.values(supportedMethods);
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  parameterName: string;
  parameterType: string;
}
