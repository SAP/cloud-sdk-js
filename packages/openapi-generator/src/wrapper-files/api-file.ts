import { codeBlock, partition, unique } from '@sap-cloud-sdk/util';
import { OpenApiDocument, OpenApiOperation } from '../openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Get the content of the SAP Cloud SDK API wrapper.
 * @param openApiDocument Parsed service.
 * @returns The generated code for the SDK API wrapper.
 */
export function apiFile(openApiDocument: OpenApiDocument): string {
  const requestBodyTypes = getRequestBodyTypes(openApiDocument);
  return codeBlock`
import { OpenApiRequestBuilder } from '@sap-cloud-sdk/core';
import { DefaultApi } from './openapi/api';
${
  requestBodyTypes
    ? `import { ${requestBodyTypes} } from './openapi/model';`
    : ''
}

export const ${openApiDocument.apiName} = {
  ${getOperations(openApiDocument)}
};
`;
}

/**
 * Get the types for all request body types in the given service.
 * @param openApiDocument Parsed service metadata.
 * @returns The list of body types as a string.
 */
function getRequestBodyTypes(openApiDocument: OpenApiDocument): string {
  const bodyTypes = openApiDocument.operations
    .map(operation => operation.requestBody?.parameterType)
    .filter(requestBody => typeof requestBody !== 'undefined') as string[];

  return unique(bodyTypes).join(', ');
}

/**
 * Get all operation representations for the given service.
 * @param openApiDocument Parsed service.
 * @returns All operations as a string.
 */
function getOperations(openApiDocument: OpenApiDocument): string {
  return openApiDocument.operations
    .map(operation => getOperation(operation))
    .join(',\n');
}

/**
 * Get the string representation of one operation.
 * @param operation Operation to serialize.
 * @returns The operation as a string.
 */
function getOperation(operation: OpenApiOperation): string {
  const params = getParams(operation);
  const argsQuestionMark = params.every(param => !param.required) ? '?' : '';
  const paramsArg = params.length
    ? codeBlock`args${argsQuestionMark}: {
  ${params
    .map(param => `${param.name}${param.required ? '' : '?'}: ${param.type}`)
    .join(',\n')}
}`
    : '';
  const requestBuilderParams = [
    'DefaultApi',
    `'${operation.operationName}'`,
    ...params.map(param => `args${argsQuestionMark}.${param.name}`)
  ];

  return codeBlock`
${
  operation.operationName
}: (${paramsArg}) => new OpenApiRequestBuilder<DefaultApi, '${
    operation.operationName
  }'>(
  ${requestBuilderParams.join(',\n')}
)`;
}

interface Parameter {
  type: string;
  name: string;
  required?: boolean;
}

function getParams(operation: OpenApiOperation): Parameter[] {
  const parameters = [
    ...operation.parameters,
    ...getRequestBodyParams(operation)
  ];

  const [required, optional] = partition(parameters, param => !!param.required);
  return [...required, ...optional];
}

function getRequestBodyParams(operation: OpenApiOperation): Parameter[] {
  return operation.requestBody
    ? [
        {
          name: operation.requestBody.parameterName,
          type: operation.requestBody.parameterType,
          required: operation.requestBody.required
        }
      ]
    : [];
}
