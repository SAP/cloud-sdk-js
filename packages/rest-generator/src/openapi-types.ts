import { $Refs } from '@apidevtools/swagger-parser';
import { OpenAPIV3 } from 'openapi-types';

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of a parsed OpenApi Service.
 */
export interface OpenApiDocument {
  apiName: string;
  serviceDirName: string;
  refs: $Refs;
  operations: OpenApiOperation[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Representation of an operation.
 */
export interface OpenApiOperation extends OpenAPIV3.OperationObject {
  method: string;
  pattern: string;
  requestBody?: OpenApiRequestBody;
  parameters: OpenApiParameter[];
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Methods supported by OpenApi and SAP Cloud SDK.
 */
export enum Method {
  get = 'get',
  put = 'put',
  post = 'post',
  patch = 'patch',
  delete = 'delete',
  head = 'head',
  options = 'options'
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenatiation of a parameter for both queries and path parameters.
 */
export interface OpenApiParameter extends OpenAPIV3.ParameterObject {
  type: string;
}

/**
 * @experimental This API is experimental and might change in newer versions. Use with caution.
 * Represenation of the request body.
 */
export interface OpenApiRequestBody extends OpenAPIV3.RequestBodyObject {
  name: string;
  type: string;
}
