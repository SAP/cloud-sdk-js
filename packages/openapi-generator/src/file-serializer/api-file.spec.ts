import { OpenApiApi, OpenApiReferenceSchema } from '../openapi-types';
import { apiDocumentation, apiFile } from './api-file';

describe('apiFile', () => {
  it('apiFile serializes api file with one operation and no references', () => {
    const api: OpenApiApi = {
      name: 'TestApi',
      operations: [
        {
          operationId: 'getFn',
          method: 'get',
          tags: [],
          pathParameters: [
            {
              in: 'path',
              name: 'id',
              originalName: 'id',
              schema: { type: 'string' },
              required: true,
              schemaProperties: {}
            }
          ],
          queryParameters: [],
          response: { type: 'any' },
          responses: { 200: { description: 'some response description' } },
          pathPattern: 'test/{id}'
        }
      ]
    };
    expect(apiFile(api, 'MyserviceName')).toMatchInlineSnapshot(`
      "import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
      /**
       * Representation of the 'TestApi'.
       * This API is part of the 'MyserviceName' service.
       */
      export const TestApi = {
        /**
         * Create a request builder for execution of get requests to the 'test/{id}' endpoint.
         * @param id - Path parameter.
         * @returns The request builder, use the \`execute()\` method to trigger the request.
         */
        getFn: (id: string) => new OpenApiRequestBuilder<any>(
          'get',
          'test/{id}',
          {
                pathParameters: { id }
              }
        )
      };"
    `);
  });

  it('apiFile serializes api file with multiple operations and references', () => {
    const api: OpenApiApi = {
      name: 'TestApi',
      operations: [
        {
          operationId: 'getFn',
          method: 'get',
          responses: { 200: { description: 'some response description' } },
          tags: [],
          pathParameters: [
            {
              in: 'path',
              name: 'id',
              originalName: 'id',
              schema: {
                $ref: '#/components/schemas/PathParameterType',
                schemaName: 'PathParameterType'
              } as OpenApiReferenceSchema,
              required: true,
              schemaProperties: {}
            }
          ],
          queryParameters: [
            {
              in: 'query',
              name: 'queryParam',
              originalName: 'queryParam',
              schema: {
                $ref: '#/components/schemas/QueryParameterType',
                schemaName: 'QueryParameterType'
              } as OpenApiReferenceSchema,
              required: true,
              schemaProperties: {}
            }
          ],
          pathPattern: 'test/{id}',
          response: { type: 'string' }
        },
        {
          operationId: 'createFn',
          method: 'post',
          responses: { 201: { description: 'some response description' } },
          tags: [],
          pathParameters: [],
          queryParameters: [],
          requestBody: {
            required: true,
            schema: {
              $ref: '#/components/schemas/RefType',
              schemaName: 'RefType'
            } as OpenApiReferenceSchema
          },
          pathPattern: 'test',
          response: {
            $ref: '#/components/schemas/ResponseType',
            schemaName: 'ResponseType'
          } as OpenApiReferenceSchema
        }
      ]
    };
    expect(apiFile(api, 'MyserviceName')).toMatchInlineSnapshot(`
      "import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
      import type { QueryParameterType, RefType, ResponseType } from './schema';
      /**
       * Representation of the 'TestApi'.
       * This API is part of the 'MyserviceName' service.
       */
      export const TestApi = {
        /**
         * Create a request builder for execution of get requests to the 'test/{id}' endpoint.
         * @param id - Path parameter.
         * @param queryParameters - Object containing the following keys: queryParam.
         * @returns The request builder, use the \`execute()\` method to trigger the request.
         */
        getFn: (id: string, queryParameters: {'queryParam': QueryParameterType}) => new OpenApiRequestBuilder<string>(
          'get',
          'test/{id}',
          {
                pathParameters: { id },
                queryParameters
              }
        ),
        /**
         * Create a request builder for execution of post requests to the 'test' endpoint.
         * @param body - Request body.
         * @returns The request builder, use the \`execute()\` method to trigger the request.
         */
        createFn: (body: RefType) => new OpenApiRequestBuilder<ResponseType>(
          'post',
          'test',
          {
                body
              }
        )
      };"
    `);
  });

  it('creates a api File with documentation', () => {
    const api: OpenApiApi = {
      name: 'TestApi',
      operations: [
        {
          operationId: 'getFn',
          method: 'get',
          responses: { 200: { description: 'some response description' } },
          tags: [],
          pathParameters: [],
          queryParameters: [],
          response: { type: 'any' },
          pathPattern: 'test'
        }
      ]
    };

    expect(apiFile(api, 'TestService')).toMatchInlineSnapshot(`
      "import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
      /**
       * Representation of the 'TestApi'.
       * This API is part of the 'TestService' service.
       */
      export const TestApi = {
        /**
         * Create a request builder for execution of get requests to the 'test' endpoint.
         * @returns The request builder, use the \`execute()\` method to trigger the request.
         */
        getFn: () => new OpenApiRequestBuilder<any>(
          'get',
          'test'
        )
      };"
    `);
  });

  it('creates documentation for the api', () => {
    expect(apiDocumentation({ name: 'TestApi' } as any, 'TestService'))
      .toMatchInlineSnapshot(`
      "/**
       * Representation of the 'TestApi'.
       * This API is part of the 'TestService' service.
       */"
    `);
  });
});
