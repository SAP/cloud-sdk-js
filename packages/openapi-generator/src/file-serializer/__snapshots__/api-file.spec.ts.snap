// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiFile apiFile serializes api file with multiple operations and references 1`] = `
"import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
import type { QueryParameterType, RefType, ResponseType } from './schema';
/**
 * Representation of the 'TestApi'.
 * This API is part of the 'MyServiceName' service.
 */
export const TestApi = {
  /**
   * Create a request builder for execution of get requests to the 'test/{id}' endpoint.
   * @param id - Path parameter.
   * @param queryParameters - Object containing the following keys: queryParam.
   * @param headerParameters - Object containing the following keys: headerParam.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  getFn: (id: string, queryParameters: {'queryParam': QueryParameterType}, headerParameters?: {'headerParam'?: string}) => new OpenApiRequestBuilder<string>(
    'get',
    "test/{id}",
    {
          pathParameters: { id },
          queryParameters,
          headerParameters
        }
  ),
  /**
   * Create a request builder for execution of post requests to the 'test' endpoint.
   * @param body - Request body.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  createFn: (body: RefType) => new OpenApiRequestBuilder<ResponseType>(
    'post',
    "test",
    {
          body
        }
  )
};"
`;

exports[`apiFile apiFile serializes api file with multiple operations and references 2`] = `
"import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
import type { QueryParameterType, RefType, ResponseType } from './schema/index.js';
/**
 * Representation of the 'TestApi'.
 * This API is part of the 'MyServiceName' service.
 */
export const TestApi = {
  /**
   * Create a request builder for execution of get requests to the 'test/{id}' endpoint.
   * @param id - Path parameter.
   * @param queryParameters - Object containing the following keys: queryParam.
   * @param headerParameters - Object containing the following keys: headerParam.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  getFn: (id: string, queryParameters: {'queryParam': QueryParameterType}, headerParameters?: {'headerParam'?: string}) => new OpenApiRequestBuilder<string>(
    'get',
    "test/{id}",
    {
          pathParameters: { id },
          queryParameters,
          headerParameters
        }
  ),
  /**
   * Create a request builder for execution of post requests to the 'test' endpoint.
   * @param body - Request body.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  createFn: (body: RefType) => new OpenApiRequestBuilder<ResponseType>(
    'post',
    "test",
    {
          body
        }
  )
};"
`;

exports[`apiFile apiFile serializes api file with one operation and no references 1`] = `
"import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
/**
 * Representation of the 'TestApi'.
 * This API is part of the 'MyServiceName' service.
 */
export const TestApi = {
  /**
   * Create a request builder for execution of get requests to the 'test/{id}' endpoint.
   * @param id - Path parameter.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  getFn: (id: string) => new OpenApiRequestBuilder<any>(
    'get',
    "test/{id}",
    {
          pathParameters: { id }
        }
  )
};"
`;

exports[`apiFile creates a api File with documentation 1`] = `
"import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
/**
 * Representation of the 'TestApi'.
 * This API is part of the 'TestService' service.
 */
export const TestApi = {
  /**
   * Create a request builder for execution of get requests to the 'test' endpoint.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  getFn: () => new OpenApiRequestBuilder<any>(
    'get',
    "test"
  )
};"
`;

exports[`apiFile creates an api file following the esm pattern 1`] = `
"import { OpenApiRequestBuilder } from '@sap-cloud-sdk/openapi';
import type { QueryParameterType, RefType, ResponseType } from './schema/index.js';
/**
 * Representation of the 'TestApi'.
 * This API is part of the 'MyServiceName' service.
 */
export const TestApi = {
  /**
   * Create a request builder for execution of get requests to the 'test/{id}' endpoint.
   * @param id - Path parameter.
   * @param queryParameters - Object containing the following keys: queryParam.
   * @param headerParameters - Object containing the following keys: headerParam.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  getFn: (id: string, queryParameters: {'queryParam': QueryParameterType}, headerParameters?: {'headerParam'?: string}) => new OpenApiRequestBuilder<string>(
    'get',
    "test/{id}",
    {
          pathParameters: { id },
          queryParameters,
          headerParameters
        }
  ),
  /**
   * Create a request builder for execution of post requests to the 'test' endpoint.
   * @param body - Request body.
   * @returns The request builder, use the \`execute()\` method to trigger the request.
   */
  createFn: (body: RefType) => new OpenApiRequestBuilder<ResponseType>(
    'post',
    "test",
    {
          body
        }
  )
};"
`;

exports[`apiFile creates documentation for the api 1`] = `
"/**
 * Representation of the 'TestApi'.
 * This API is part of the 'TestService' service.
 */"
`;
